First stab at event loopery

wlonk opened the pull request:
This should probably farm out the rendering into its own thread, the input listening into another, and perhaps another for background processes?

The real question, of course, is what the API will be like. For now, we seem to be going the subclass-and-add-custom-methods route.

wlonk committed:
First stab at event loopery

This should probably farm out the rendering into its own thread, the
input listening into another, and perhaps another for background
processes?

The real question, of course, is what the API will be like. For now, we
seem to be going the subclass-and-add-custom-methods route.

ojacobson commented on hexes/hexes.py:
```diff
@@ -51,6 +54,26 @@ def __repr__(self):
+
+    @text.setter
+    def text(self, val):
+        dirty = val != self._text
```
`self._text, self.dirty = modify(self._text, val)`

``` python
def modify(original, new):
    return new, (original == new)
```

ojacobson replied:
... Which is then a candidate for a descriptor of its own, at which point: exeunt boilerplate

wlonk replied:
oh, that's a nice pattern.

ojacobson replied:
In fact the computation of `dirty` needs to be smarter than that anyways, so again, a descriptor seems warranted. Otherwise you get shit like

``` python
a.text = "foo"
assert a.dirty
a.text = "foo"
assert not a.dirty
```

wlonk replied:
Er, isn't that desirable? If it hasn't changed, not dirty—not if it hasn't been assigned to, not dirty.

wlonk committed:
Addressing a comment from @ojacobson

There's still a lot of repetitive code here, but this is an interesting
approach to minimizing it.

wlonk commented:
Conversation in IRC clarified @ojacobson's comment: writing the same value twice between renderings shouldn't mark the value as clean. Rendering should.

As a side-effect of the dirty-the-whole-tree that I have here, I wouldn't easily notice this, but it is worth correcting.

wlonk committed:
Adds events and listeners

Attach listeners to an app object with a decorator, @app.on("event").
